zkvm/src/predicate.rs

```diff
@@ -27,19 +28,39 @@ pub enum Predicate {
     /// constructing aggregated signature.
     Key(VerificationKey),
 
-    /// Representation of a predicate as commitment to a program and blinding factor.
-    Program(Program, Vec<u8>),
-
-    /// Disjunction of n predicates.
-    Or(PredicateDisjunction),
+    /// Taproot Merkle tree.
+    Tree(PredicateTree),
 }
 
+/// Represents a ZkVM predicate tree.
 #[derive(Clone, Debug)]
-pub struct PredicateDisjunction {
-    preds: Vec<Predicate>,
-    // We precompute the disjunction predicate when composing it via `Predicate::or()`,
+pub struct PredicateTree {
+    /// Vector of the programs and blinding factors, stored as Merkelized predicate leafs.
+    pub leaves: Vec<PredicateLeaf>,
+    /// Verification key for the tree.
+    pub key: VerificationKey,
+    blinding_key: [u8; 32],
+    // We precompute the tweaked aggregated signing key when composing it via `Predicate::tree()`,
     // so that we can keep `to_point`/`encode` methods non-failable across all types.
     precomputed_point: CompressedRistretto,
+
+    // Scalar that's added to the signing key - a hash of the merkle root and the pubkey
+    adjustment_factor: Scalar,
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct CallProof {
+    // Tweaked verification key
```

Comment:
Why is this tweaked? In the context of `call`, this key is just a piece of extra data necessary to verify a commitment. And it's supposed to be the pure verification key X in the formula `P = X + hash(X,Root)*B`, not the predicate point itself (P).

Reply:
Good catch -- I think this is just a hold-over comment from a previous incorrect implementation. From the construction of the `PointOp` [here](https://github.com/interstellar/slingshot/pull/268/files/abe92d6b5cf4c4944d10ba4a12920edeb5cf9bff#diff-7e106625822af74f4132c28bcb9fe28bR128), this is the pure verification key `X`.